### PR TITLE
Make XR/VR presenting code async

### PR DESCRIPTION
**Pull request by Manishearth**
Previously we only pretended to be async, we returned promises that were already resolved because we synchronously blocked on the channel.

This:

 - Moves XR presentation to session creation (where it belongs, see https://github.com/immersive-web/webxr/issues/453)
 - Factors out common presentation code in a way that supports asynchronously resolving promises
 - Uses this for `VRDisplay::RequestPresent()` and `XR::RequestSession()`

r? @jdm

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22649)

### CODE REVIEW

**Comment by Manishearth** [comment]
I haven't actually tested this, it _may_ blow up because the vr code doesn't expect an empty context

**Comment by Manishearth** [comment]
Oh, it definitely will, we have some unwraps there.

**Comment by Manishearth** [comment]
And the spec hasn't really ... decided what to do with empty contexts.

(now I remember why I didn't do this in the first place)

The RAF thread complicates this, but I might be able to do something relatively obvious until they decide the behavior.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/vrdisplay.rs [hunk 1]
```diff
@@ -464,6 +437,59 @@ impl VRDisplay {
...
+            .unwrap();
+        let promise = TrustedPromise::new(promise);
+        let this = Trusted::new(self);
+        let ctx = ctx.map(|c| Trusted::new(c));
+        let global = self.global();
+        let window = global.as_window();
+        let (task_source, canceller) = window
+            .task_manager()
+            .dom_manipulation_task_source_with_canceller();
+        thread::spawn(move || {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/script/dom/vrdisplay.rs]: Oh, the recv() is supposed to be outside the task, my bad.

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/script/dom/vrdisplay.rs]: Oh, the recv() is supposed to be outside the task, my bad.

**Review by jdm — APPROVED** [review]

**Comment by Manishearth** [comment]
oh this wasn't supposed to merge, like I said this will panic unconditionally

(and that's a bit tricky to fix)

I've got a WIP for fixing this.

Since I'm the only one using XR rn it's fine to leave it as is though

**Comment by jdm** [comment]
Oops!

**Pull request merged**